File: imagepicker/src/main/java/com/nguyenhoanglam/imagepicker/ui/imagepicker/ImagePickerActivity.java
package com.nguyenhoanglam.imagepicker.ui.imagepicker;

import android.Manifest;
import android.content.Intent;
import android.content.res.Configuration;
import android.database.ContentObserver;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Parcelable;
import android.provider.MediaStore;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Toast;

import com.nguyenhoanglam.imagepicker.R;
import com.nguyenhoanglam.imagepicker.helper.CameraHelper;
import com.nguyenhoanglam.imagepicker.helper.LogHelper;
import com.nguyenhoanglam.imagepicker.helper.PermissionHelper;
import com.nguyenhoanglam.imagepicker.listener.OnBackAction;
import com.nguyenhoanglam.imagepicker.listener.OnFolderClickListener;
import com.nguyenhoanglam.imagepicker.listener.OnImageClickListener;
import com.nguyenhoanglam.imagepicker.listener.OnImageSelectionListener;
import com.nguyenhoanglam.imagepicker.model.Config;
import com.nguyenhoanglam.imagepicker.model.Folder;
import com.nguyenhoanglam.imagepicker.model.Image;
import com.nguyenhoanglam.imagepicker.widget.ImagePickerToolbar;
import com.nguyenhoanglam.imagepicker.widget.ProgressWheel;
import com.nguyenhoanglam.imagepicker.widget.SnackBarView;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by hoanglam on 7/31/16.
 */
public class ImagePickerActivity extends AppCompatActivity implements ImagePickerView {

    private ImagePickerToolbar toolbar;
    private RecyclerViewManager recyclerViewManager;
    private RecyclerView recyclerView;
    private ProgressWheel progressWheel;
    private View emptyLayout;
    private SnackBarView snackBar;

    private Config config;
    private Handler handler;
    private ContentObserver observer;
    private ImagePickerPresenter presenter;
    private LogHelper logger = LogHelper.getInstance();


    private OnImageClickListener imageClickListener = new OnImageClickListener() {
        @Override
        public boolean onImageClick(View view, int position, boolean isSelected) {
            return recyclerViewManager.selectImage();
        }
    };

    private OnFolderClickListener folderClickListener = new OnFolderClickListener() {
        @Override
        public void onFolderClick(Folder folder) {
            setImageAdapter(folder.getImages(), folder.getFolderName());
        }
    };

    private View.OnClickListener backClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            onBackPressed();
        }
    };

    private View.OnClickListener cameraClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            captureImageWithPermission();
        }
    };

    private View.OnClickListener doneClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            onDone();
        }
    };


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Intent intent = getIntent();
        if (intent == null) {
            finish();
            return;
        }

        config = intent.getParcelableExtra(Config.EXTRA_CONFIG);
        if (config.isKeepScreenOn()) {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
        }

        setContentView(R.layout.imagepicker_activity_picker);

        setupView();
        setupComponents();
        setupToolbar();

    }

    private void setupView() {
        toolbar = findViewById(R.id.toolbar);
        recyclerView = findViewById(R.id.recyclerView);
        progressWheel = findViewById(R.id.progressWheel);
        emptyLayout = findViewById(R.id.layout_empty);
        snackBar = findViewById(R.id.snackbar);

        Window window = getWindow();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            window.setStatusBarColor(config.getStatusBarColor());
        }

        progressWheel.setBarColor(config.getProgressBarColor());
        findViewById(R.id.container).setBackgroundColor(config.getBackgroundColor());


    }

    private void setupComponents() {
        recyclerViewManager = new RecyclerViewManager(recyclerView, config, getResources().getConfiguration().orientation);
        recyclerViewManager.setupAdapters(imageClickListener, folderClickListener);
        recyclerViewManager.setOnImageSelectionListener(new OnImageSelectionListener() {
            @Override
            public void onSelectionUpdate(List<Image> images) {
                invalidateToolbar();
                if (!config.isMultipleMode() && !images.isEmpty()) {
                    onDone();
                }
            }
        });

        presenter = new ImagePickerPresenter(new ImageFileLoader(this));
        presenter.attachView(this);
    }

    private void setupToolbar() {
        toolbar.config(config);
        toolbar.setOnBackClickListener(backClickListener);
        toolbar.setOnCameraClickListener(cameraClickListener);
        toolbar.setOnDoneClickListener(doneClickListener);
    }

    @Override
    protected void onResume() {
        super.onResume();
        getDataWithPermission();
    }


    private void setImageAdapter(List<Image> images, String title) {
        recyclerViewManager.setImageAdapter(images, title);
        invalidateToolbar();
    }

    private void setFolderAdapter(List<Folder> folders) {
        recyclerViewManager.setFolderAdapter(folders);
        invalidateToolbar();
    }

    private void invalidateToolbar() {
        toolbar.setTitle(recyclerViewManager.getTitle());
        toolbar.showDoneButton(recyclerViewManager.isShowDoneButton());

    }

    private void onDone() {
        presenter.onDoneSelectImages(recyclerViewManager.getSelectedImages());
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        recyclerViewManager.changeOrientation(newConfig.orientation);
    }


    private void getDataWithPermission() {

        final String[] permissions = new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE};

        PermissionHelper.checkPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE, new PermissionHelper.PermissionAskListener() {
            @Override
            public void onNeedPermission() {
                PermissionHelper.requestAllPermissions(ImagePickerActivity.this, permissions, Config.RC_WRITE_EXTERNAL_STORAGE_PERMISSION);
            }

            @Override
            public void onPermissionPreviouslyDenied() {
                PermissionHelper.requestAllPermissions(ImagePickerActivity.this, permissions, Config.RC_WRITE_EXTERNAL_STORAGE_PERMISSION);
            }

            @Override
            public void onPermissionDisabled() {
                snackBar.show(R.string.imagepicker_msg_no_write_external_storage_permission, new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        PermissionHelper.openAppSettings(ImagePickerActivity.this);
                    }
                });
            }

            @Override
            public void onPermissionGranted() {
                getData();
            }
        });
    }

    private void getData() {
        presenter.abortLoading();
        presenter.loadImages(config.isFolderMode());
    }


    private void captureImageWithPermission() {

        final String[] permissions = new String[]{Manifest.permission.CAMERA};

        PermissionHelper.checkPermission(this, Manifest.permission.CAMERA, new PermissionHelper.PermissionAskListener() {
            @Override
            public void onNeedPermission() {
                PermissionHelper.requestAllPermissions(ImagePickerActivity.this, permissions, Config.RC_CAMERA_PERMISSION);
            }

            @Override
            public void onPermissionPreviouslyDenied() {
                PermissionHelper.requestAllPermissions(ImagePickerActivity.this, permissions, Config.RC_CAMERA_PERMISSION);
            }

            @Override
            public void onPermissionDisabled() {
                snackBar.show(R.string.imagepicker_msg_no_camera_permission, new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        PermissionHelper.openAppSettings(ImagePickerActivity.this);
                    }
                });
            }

            @Override
            public void onPermissionGranted() {
                captureImage();
            }
        });
    }


    private void captureImage() {
        if (!CameraHelper.checkCameraAvailability(this)) {
            return;
        }
        presenter.captureImage(this, config, Config.RC_CAPTURE_IMAGE);
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == Config.RC_CAPTURE_IMAGE && resultCode == RESULT_OK) {
            presenter.finishCaptureImage(this, data, config);
        }
    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {

        switch (requestCode) {
            case Config.RC_WRITE_EXTERNAL_STORAGE_PERMISSION: {
                if (PermissionHelper.hasGranted(grantResults)) {
                    logger.d("Write External permission granted");
                    getData();
                    return;
                }
                logger.e("Permission not granted: results len = " + grantResults.length +
                        " Result code = " + (grantResults.length > 0 ? grantResults[0] : "(empty)"));
                finish();
            }
            case Config.RC_CAMERA_PERMISSION: {
                if (PermissionHelper.hasGranted(grantResults)) {
                    logger.d("Camera permission granted");
                    captureImage();
                    return;
                }
                logger.e("Permission not granted: results len = " + grantResults.length +
                        " Result code = " + (grantResults.length > 0 ? grantResults[0] : "(empty)"));
                break;
            }
            default: {
                logger.d("Got unexpected permission result: " + requestCode);
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
                break;
            }
        }
    }

    @Override
    protected void onStart() {
        super.onStart();

        if (handler == null) {
            handler = new Handler();
        }
        observer = new ContentObserver(handler) {
            @Override
            public void onChange(boolean selfChange) {
                getData();
            }
        };
        getContentResolver().registerContentObserver(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, false, observer);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (presenter != null) {
            presenter.abortLoading();
            presenter.detachView();
        }

        if (observer != null) {
            getContentResolver().unregisterContentObserver(observer);
            observer = null;
        }

        if (handler != null) {
            handler.removeCallbacksAndMessages(null);
            handler = null;
        }
    }

    @Override
    public void onBackPressed() {
        recyclerViewManager.handleBack(new OnBackAction() {
            @Override
            public void onBackToFolder() {
                invalidateToolbar();
            }

            @Override
            public void onFinishImagePicker() {
                setResult(RESULT_CANCELED);
                finish();
            }
        });
    }

    /**
     * MVP view methods
     */

    @Override
    public void showLoading(boolean isLoading) {
        progressWheel.setVisibility(isLoading ? View.VISIBLE : View.GONE);
        recyclerView.setVisibility(isLoading ? View.GONE : View.VISIBLE);
        emptyLayout.setVisibility(View.GONE);
    }

    @Override
    public void showFetchCompleted(List<Image> images, List<Folder> folders) {
        if (config.isFolderMode()) {
            setFolderAdapter(folders);
        } else {
            setImageAdapter(images, config.getImageTitle());
        }
    }

    @Override
    public void showError(Throwable throwable) {
        String message = getString(R.string.imagepicker_error_unknown);
        if (throwable != null && throwable instanceof NullPointerException) {
            message = getString(R.string.imagepicker_error_images_not_exist);
        }
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showEmpty() {
        progressWheel.setVisibility(View.GONE);
        recyclerView.setVisibility(View.GONE);
        emptyLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void showCapturedImage(List<Image> images) {
        boolean shouldSelect = recyclerViewManager.selectImage();
        if (shouldSelect) {
            recyclerViewManager.addSelectedImages(images);
        }
        getDataWithPermission();
    }

    @Override
    public void finishPickImages(List<Image> images) {
        Intent data = new Intent();
        data.putParcelableArrayListExtra(Config.EXTRA_IMAGES, (ArrayList<? extends Parcelable>) images);
        setResult(RESULT_OK, data);
        finish();
    }
}

File: imagepicker/src/main/java/com/nguyenhoanglam/imagepicker/ui/imagepicker/ImagePickerActivity.kt
/*
 * Copyright (C) 2023 Image Picker
 * Author: Nguyen Hoang Lam <hoanglamvn90@gmail.com>
 */

package com.nguyenhoanglam.imagepicker.ui.imagepicker

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.View
import androidx.activity.OnBackPressedCallback
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.nguyenhoanglam.imagepicker.R
import com.nguyenhoanglam.imagepicker.databinding.ImagepickerActivityImagepickerBinding
import com.nguyenhoanglam.imagepicker.helper.Constants
import com.nguyenhoanglam.imagepicker.helper.DeviceHelper
import com.nguyenhoanglam.imagepicker.helper.GlideHelper
import com.nguyenhoanglam.imagepicker.helper.ImageHelper
import com.nguyenhoanglam.imagepicker.helper.PermissionHelper
import com.nguyenhoanglam.imagepicker.helper.SelectionState
import com.nguyenhoanglam.imagepicker.helper.ToastHelper
import com.nguyenhoanglam.imagepicker.listener.OnFolderClickListener
import com.nguyenhoanglam.imagepicker.listener.OnImageSelectListener
import com.nguyenhoanglam.imagepicker.model.Folder
import com.nguyenhoanglam.imagepicker.model.Image
import com.nguyenhoanglam.imagepicker.model.ImagePickerConfig
import com.nguyenhoanglam.imagepicker.model.StatusBarContent
import com.nguyenhoanglam.imagepicker.ui.camera.CameraModule
import com.nguyenhoanglam.imagepicker.ui.camera.OnImageReadyListener

class ImagePickerActivity : AppCompatActivity(), OnFolderClickListener, OnImageSelectListener {

    private lateinit var binding: ImagepickerActivityImagepickerBinding
    private lateinit var config: ImagePickerConfig

    private lateinit var viewModel: ImagePickerViewModel
    private val cameraModule = CameraModule()

    private val backClickListener = View.OnClickListener { handleBackPress() }
    private val cameraClickListener = View.OnClickListener { captureImageWithPermission() }
    private val selectAllClickListener = View.OnClickListener { handleSelectAllImages() }
    private val unselectAllClickListener = View.OnClickListener { handleUnselectAllImages() }
    private val doneClickListener = View.OnClickListener { onDone() }

    private val resultLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                cameraModule.saveImage(this@ImagePickerActivity,
                    config,
                    object : OnImageReadyListener {
                        override fun onImageReady(images: ArrayList<Image>) {
                            fetchDataWithPermission()
                        }

                        override fun onImageNotReady() {
                            fetchDataWithPermission()
                        }
                    })
            }
        }

    @Suppress("DEPRECATION")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (intent == null) {
            finish()
            return
        }

        config = if (DeviceHelper.isMinSdk33) intent.getParcelableExtra(
            Constants.EXTRA_CONFIG, ImagePickerConfig::class.java
        )!! else intent.getParcelableExtra(Constants.EXTRA_CONFIG)!!
        config.initDefaultValues(this@ImagePickerActivity)

        // Setup status bar theme
        window.statusBarColor = Color.parseColor(config.customColor!!.statusBar)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        WindowInsetsControllerCompat(window, window.decorView).isAppearanceLightStatusBars =
            config.statusBarContentMode == StatusBarContent.DARK

        GlideHelper.setConfig(
            config.isImageTransitionEnabled,
            config.customDrawable!!.loadingImagePlaceholder,
            config.customDrawable!!.errorImagePlaceholder
        )

        binding = ImagepickerActivityImagepickerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(
            this, ImagePickerViewModelFactory(this.application)
        )[ImagePickerViewModel::class.java]
        viewModel.apply {
            setConfig(config)
            result.observe(this@ImagePickerActivity) {
                val fragment = getCurrentFragment()
                if (fragment != null && fragment is ImageFragment) {
                    configSelectAllButtons(
                        fragment.getBucketId(), it.images, viewModel.selectedImages.value
                    )
                }
            }
            selectedImages.observe(this@ImagePickerActivity) {
                binding.toolbar.showDoneButton(config.isAlwaysShowDoneButton || it.isNotEmpty())

                val fragment = getCurrentFragment()
                if (fragment != null && fragment is ImageFragment) {
                    configSelectAllButtons(
                        fragment.getBucketId(), viewModel.result.value?.images, it
                    )
                }
            }
        }

        setupViews()

        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                handleBackPress()
            }
        })
    }

    override fun onResume() {
        super.onResume()
        fetchDataWithPermission()
    }

    private fun setupViews() {
        binding.toolbar.apply {
            setConfig(config)
            setOnBackClickListener(backClickListener)
            setOnCameraClickListener(cameraClickListener)
            setOnSelectAllClickListener(selectAllClickListener)
            setOnUnselectAllClickListener(unselectAllClickListener)
            setOnDoneClickListener(doneClickListener)
        }

        binding.snackbar.config(config)

        val initialFragment =
            if (config.isFolderMode) FolderFragment.newInstance(config.folderGridCount)
            else ImageFragment.newInstance(config.imageGridCount)

        supportFragmentManager.beginTransaction().replace(R.id.fragmentContainer, initialFragment)
            .commit()
    }

    private fun getCurrentFragment(): Fragment? =
        supportFragmentManager.findFragmentById(R.id.fragmentContainer)


    private fun isFolderFragment(): Boolean {
        val fragment = getCurrentFragment()
        return fragment != null && fragment is FolderFragment
    }

    private fun handleBackPress() {
        if (supportFragmentManager.backStackEntryCount > 0) {
            supportFragmentManager.popBackStackImmediate()
            if (isFolderFragment()) {
                binding.toolbar.setTitle(config.folderTitle)
                binding.toolbar.hideSelectButtons()
            }
        } else {
            finishPickImages(config.selectedImages)
        }
    }

    private fun fetchDataWithPermission() {
        val readPermission =
            if (DeviceHelper.isMinSdk33) Manifest.permission.READ_MEDIA_IMAGES else Manifest.permission.READ_EXTERNAL_STORAGE

        when (PermissionHelper.checkPermission(this@ImagePickerActivity, readPermission)) {
            PermissionHelper.STATUS.GRANTED -> fetchData()

            PermissionHelper.STATUS.NOT_GRANTED -> PermissionHelper.requestAllPermissions(
                this@ImagePickerActivity, arrayOf(readPermission), Constants.RC_READ_PERMISSION
            )

            PermissionHelper.STATUS.DENIED -> PermissionHelper.requestAllPermissions(
                this@ImagePickerActivity, arrayOf(readPermission), Constants.RC_READ_PERMISSION
            )

            else -> binding.snackbar.show(
                config.customMessage?.noPhotoAccessPermission
            ) {
                PermissionHelper.openAppSettings(this@ImagePickerActivity)
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        when (requestCode) {
            Constants.RC_READ_PERMISSION -> {
                if (PermissionHelper.hasGranted(grantResults)) {
                    fetchData()
                } else {
                    finish()
                }
            }

            Constants.RC_WRITE_PERMISSION, Constants.RC_CAMERA_PERMISSION -> {
                if (PermissionHelper.hasGranted(grantResults)) {
                    captureImage()
                }
            }

            else -> {
                super.onRequestPermissionsResult(requestCode, permissions, grantResults)
            }
        }
    }

    private fun fetchData() = viewModel.fetchImages()

    private fun onDone() {
        val selectedImages = viewModel.selectedImages.value
        finishPickImages(selectedImages ?: arrayListOf())
    }

    private fun handleSelectAllImages() {
        val fragment = getCurrentFragment()
        if (fragment != null && fragment is ImageFragment) {
            fragment.selectAllImages()
        }
    }

    private fun handleUnselectAllImages() {
        val fragment = getCurrentFragment()
        if (fragment != null && fragment is ImageFragment) {
            fragment.unselectAllImages()
        }
    }

    private fun configSelectAllButtons(
        bucketId: Long?, images: ArrayList<Image>?, selectedImages: ArrayList<Image>?
    ) {
        if (!config.isMultiSelectMode) return

        val state = ImageHelper.getBucketSelectionState(
            images, selectedImages, bucketId
        )

        when (state) {
            SelectionState.SELECTED -> {
                if (config.isUnselectAllEnabled) {
                    binding.toolbar.hideSelectAllButton()
                    binding.toolbar.showUnselectAllButton()
                } else {
                    binding.toolbar.hideSelectButtons()
                }
            }

            else -> {
                val imagesByBucketCount = ImageHelper.filter(images, bucketId).size
                val selectedImagesCount = selectedImages?.size ?: 0

                if (config.isSelectAllEnabled && config.limitSize >= selectedImagesCount + imagesByBucketCount) {
                    binding.toolbar.showSelectAllButton()
                    binding.toolbar.hideUnselectAllButton()
                } else {
                    binding.toolbar.hideSelectButtons()
                }
            }
        }
    }

    private fun captureImageWithPermission() {
        val isCameraPermissionDeclared = PermissionHelper.isPermissionDeclared(
            this@ImagePickerActivity, Manifest.permission.CAMERA
        )

        if (DeviceHelper.isMinSdk29) {
            if (isCameraPermissionDeclared) {
                val cameraPermission = Manifest.permission.CAMERA
                when (PermissionHelper.checkPermission(
                    this@ImagePickerActivity, cameraPermission
                )) {
                    PermissionHelper.STATUS.GRANTED -> captureImage()

                    PermissionHelper.STATUS.NOT_GRANTED -> PermissionHelper.requestAllPermissions(
                        this@ImagePickerActivity,
                        arrayOf(cameraPermission),
                        Constants.RC_CAMERA_PERMISSION
                    )

                    PermissionHelper.STATUS.DENIED -> PermissionHelper.requestAllPermissions(
                        this@ImagePickerActivity,
                        arrayOf(cameraPermission),
                        Constants.RC_CAMERA_PERMISSION
                    )

                    else -> binding.snackbar.show(
                        config.customMessage!!.noCameraPermission
                    ) {
                        PermissionHelper.openAppSettings(this@ImagePickerActivity)
                    }
                }
            } else {
                captureImage()
            }
        } else {
            if (isCameraPermissionDeclared) {
                val statuses = PermissionHelper.checkPermissions(
                    this@ImagePickerActivity,
                    arrayOf(Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE)
                )

                if (statuses[0] == PermissionHelper.STATUS.GRANTED && statuses[1] == PermissionHelper.STATUS.GRANTED) {
                    captureImage()
                } else if (statuses[0] == PermissionHelper.STATUS.DISABLED && statuses[1] == PermissionHelper.STATUS.DISABLED) {
                    binding.snackbar.show(
                        config.customMessage?.noCameraPermission
                    ) {
                        PermissionHelper.openAppSettings(this@ImagePickerActivity)
                    }
                } else if (statuses[0] == PermissionHelper.STATUS.DISABLED) {
                    binding.snackbar.show(
                        config.customMessage?.noCameraPermission
                    ) {
                        PermissionHelper.openAppSettings(this@ImagePickerActivity)
                    }
                } else if (statuses[1] == PermissionHelper.STATUS.DISABLED) {
                    binding.snackbar.show(
                        config.customMessage?.noPhotoAccessPermission
                    ) {
                        PermissionHelper.openAppSettings(this@ImagePickerActivity)
                    }
                } else {
                    val requestPermissions = ArrayList<String>()
                    for ((index, value) in statuses.withIndex()) {
                        if (value == PermissionHelper.STATUS.NOT_GRANTED || value == PermissionHelper.STATUS.DENIED) {
                            requestPermissions.add(if (index == 0) Manifest.permission.CAMERA else Manifest.permission.WRITE_EXTERNAL_STORAGE)
                        }
                    }

                    PermissionHelper.requestAllPermissions(
                        this@ImagePickerActivity,
                        requestPermissions.toTypedArray(),
                        Constants.RC_CAMERA_PERMISSION
                    )
                }
            } else {
                val writePermission = Manifest.permission.WRITE_EXTERNAL_STORAGE
                when (PermissionHelper.checkPermission(
                    this,
                    writePermission,
                )) {
                    PermissionHelper.STATUS.GRANTED -> captureImage()

                    PermissionHelper.STATUS.NOT_GRANTED -> PermissionHelper.requestAllPermissions(
                        this@ImagePickerActivity,
                        arrayOf(writePermission),
                        Constants.RC_WRITE_PERMISSION
                    )

                    PermissionHelper.STATUS.DENIED -> PermissionHelper.requestAllPermissions(
                        this@ImagePickerActivity,
                        arrayOf(writePermission),
                        Constants.RC_WRITE_PERMISSION
                    )

                    else -> binding.snackbar.show(
                        config.customMessage?.noPhotoAccessPermission
                    ) {
                        PermissionHelper.openAppSettings(this@ImagePickerActivity)
                    }
                }
            }
        }
    }


    private fun captureImage() {
        if (!DeviceHelper.checkCameraAvailability(this)) {
            return
        }

        val intent = cameraModule.getCameraIntent(this@ImagePickerActivity, config)
        if (intent == null) {
            ToastHelper.show(this, getString(R.string.imagepicker_error_camera))
            return
        }

        resultLauncher.launch(intent)
    }

    private fun finishPickImages(images: ArrayList<Image>) {
        val data = Intent()
        data.putParcelableArrayListExtra(Constants.EXTRA_IMAGES, images)
        setResult(Activity.RESULT_OK, data)
        finish()
    }


    override fun onFolderClick(folder: Folder) {
        supportFragmentManager.beginTransaction().add(
            R.id.fragmentContainer,
            ImageFragment.newInstance(folder.bucketId, config.imageGridCount)
        ).addToBackStack(null).commit()

        binding.toolbar.setTitle(folder.name)
        configSelectAllButtons(
            folder.bucketId, viewModel.result.value?.images, viewModel.selectedImages.value
        )
    }

    override fun onSelectedImagesChanged(selectedImages: ArrayList<Image>) {
        viewModel.selectedImages.value = selectedImages
    }

    override fun onSingleModeImageSelected(image: Image) {
        finishPickImages(ImageHelper.singleListFromImage(image))
    }
}